### PR TITLE
Support CWL workflow restarts (resolves #1241, resolves #1100)

### LINKED ISSUE #1100: Option --restart produces error when running CWL pipeline
``` bash
> cwltoil --restart --jobStore file:jobStore ppeline.cwl job.yml
pc: 2016-08-02 10:24:27,230 INFO: toil.lib.bioio: Logging set at level: INFO
pc: 2016-08-02 10:24:44,319 INFO: toil.lib.bioio: Logging set at level: INFO
pc: 2016-08-02 10:24:44,654 INFO: toil.jobStores.fileJobStore: Path to job store directory is '/data/cwl-test/jobStore'.
Traceback (most recent call last):
  File "/data/cwl-test/toil/bin/cwltoil", line 11, in <module>
    sys.exit(main())
  File "/data/cwl-test/toil/lib/python2.7/site-packages/toil/cwl/cwltoil.py", line 595, in main
    outobj = toil.start(wf1)
  File "/data/cwl-test/toil/lib/python2.7/site-packages/toil/common.py", line 536, in start
    raise ToilRestartException('A Toil workflow can only be started once. Use '
toil.common.ToilRestartException: A Toil workflow can only be started once. Use Toil.restart() to resume it.
```

Is the restart feature implemented for CWL?

Thanks!

### LINKED ISSUE #1241: `cwltoil --restart` does not work

### PR DESCRIPTION
**Pull request by chapmanb**
Continues #1025

Support restarting a cwltoil workflow with the `--restart` flag,
updates work by @tetron in #1025 and fixes #1100. Adds a test
to demonstrate restart capability by manipulating paths to cause the
initial run of the revsort workflow to fail.

Still a work in progress as the test fails consistently when restarting.
I could use help debugging what I'm missing to correctly restart to
finish this. If you run the test:

```
make test tests=src/toil/test/cwl/cwlTest.py::CWLTest::test_restart
```

It completes the expected failure step, but the restart fails with an
EOFError when unpickling a job store (similar to #676 and #1039):

```
INFO toil.jobStores.fileJobStore: Path to job store directory is '/tmp/user/1001/toil-test-toil.test.cwl.cwlTest.CWLTest-test_restart-l0vC5G/jobStore'.
INFO toil.common: Using the single machine batch system
WARNING toil.batchSystems.singleMachine: Limiting maxCores to CPU count of system (4).
WARNING toil.batchSystems.singleMachine: Limiting maxMemory to physically available memory (7916355584).
INFO toil.common: Created the workflow directory at /tmp/user/1001/toil-c7c91877-c1bd-4a6e-b2c8-b23d0c3d90b7
WARNING toil.batchSystems.singleMachine: Limiting maxDisk to physically available disk (15848153088).
INFO toil.batchSystems.singleMachine: Setting up the thread pool with 40 workers, given a minimum CPU fraction of 0.100000 and a maximum CPU value of 4.
INFO toil.common: Written the environment for the jobs to the environment file
INFO toil.common: Caching all jobs in job store
INFO toil.common: 4 jobs downloaded.
INFO toil.jobStores.abstractJobStore: Checking job graph connectivity...
INFO toil.jobStores.abstractJobStore: 4 jobs reachable from root.
CRITICAL toil.jobStores.abstractJobStore: Repairing job: N/g/jobLYOvLb
INFO toil.jobStores.abstractJobStore: Discarding old statistics and logs...
INFO toil.jobStores.abstractJobStore: Job store is clean
INFO toil.realtimeLogger: Real-time logging disabled
INFO toil.leader: (Re)building internal scheduler state
INFO toil.leader: Checked batch system has no running jobs and no updated jobs
INFO toil.leader: Found 2 jobs to start and 1 jobs with successors to run
INFO toil.leader: Starting the main loop
INFO toil.batchSystems.singleMachine: Executing command: '/home/chapmanb/bio/toil/venv/bin/_toil_worker /tmp/user/1001/toil-test-toil.test.cwl.cwlTest.CWLTest-test_restart-l0vC5G/jobStore N/g/jobLYOvLb'.
INFO toil.batchSystems.singleMachine: Executing command: '/home/chapmanb/bio/toil/venv/bin/_toil_worker /tmp/user/1001/toil-test-toil.test.cwl.cwlTest.CWLTest-test_restart-l0vC5G/jobStore N/2/jobv9Osdj'.
toil.leader: The jobWrapper seems to have left a log file, indicating failure: N/2/jobv9Osdj
toil.leader: Reporting file: N/2/jobv9Osdj
toil.leader: N/2/jobv9Osdj:   ---TOIL WORKER OUTPUT LOG---
toil.leader: N/2/jobv9Osdj:   INFO:rdflib:RDFLib Version: 4.2.1
toil.leader: N/2/jobv9Osdj:   INFO:toil.jobStores.fileJobStore:Path to job store directory is '/tmp/user/1001/toil-test-toil.test.cwl.cwlTest.CWLTest-test_restart-l0vC5G/jobStore'.
toil.leader: N/2/jobv9Osdj:   Traceback (most recent call last):
toil.leader: N/2/jobv9Osdj:     File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/toil-3.5.0a1-py2.7.egg/toil/worker.py", line 316, in main
toil.leader: N/2/jobv9Osdj:       job = Job._loadJob(jobWrapper.command, jobStore)
toil.leader: N/2/jobv9Osdj:     File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/toil-3.5.0a1-py2.7.egg/toil/job.py", line 673, in _loadJob
toil.leader: N/2/jobv9Osdj:       return cls._unpickle(userModule, fileHandle)
toil.leader: N/2/jobv9Osdj:     File "/home/chapmanb/bio/toil/venv/local/lib/python2.7/site-packages/toil-3.5.0a1-py2.7.egg/toil/job.py", line 696, in _unpickle
toil.leader: N/2/jobv9Osdj:       return unpickler.load()
toil.leader: N/2/jobv9Osdj:   EOFError
toil.leader: N/2/jobv9Osdj:   Exiting the worker because of a failed jobWrapper on host kunkel
toil.leader: N/2/jobv9Osdj:   ERROR:toil.worker:Exiting the worker because of a failed jobWrapper on host kunkel
toil.leader: N/2/jobv9Osdj:   WARNING:toil.jobWrapper:Due to failure we are reducing the remaining retry count of job N/2/jobv9Osdj to 0
toil.leader: N/2/jobv9Osdj:   WARNING:toil.jobWrapper:We have increased the default memory of the failed job to 2147483648 bytes
toil.leader: Job: N/2/jobv9Osdj is completely failed
```

Thanks for any tips and ideas about how best to debug and proceed.

### CODE REVIEW

**Comment by arkal** [comment]
Jenkins, test this please.

**Comment by benedictpaten** [comment]
Jenkins, test please

**Comment by benedictpaten** [comment]
Jenkins, test this please.

**Comment by cket** [comment]
The only failing test is CWL test_restart, you can see the test output here http://pastebin.com/UPqgjzLC.
It looks like these are the most relevant lines:
```
WARNING:toil.leader:X/y/jobxPa_0d:  INFO:cwltool:[job revtool.cwl] /mnt/ephemeral/tmp/tmpRYIG2q$ rev \
WARNING:toil.leader:X/y/jobxPa_0d:      /mnt/ephemeral/tmp/tmpCYsUeI/stg49ff26ed-3fbf-41ba-85bc-eac77910a205/whale.txt > /mnt/ephemeral/tmp/tmpRYIG2q/output.txt
WARNING:toil.leader:X/y/jobxPa_0d:  'rev' not found
WARNING:toil.leader:X/y/jobxPa_0d:  ERROR:cwltool:'rev' not found
```

**Comment by chapmanb** [comment]
Christopher -- thanks for looking at this one. That's right on, that's the new test I'm adding for validating restarts and I can't quite get things working for the restart so it still fails. The initial message in the PR has more details about where I'm stuck at -- the restart complains when unpickling a file and I'm not sure what needs to happen to avoid that problem and restart cleanly. Thanks for any pointers/ideas.

**Comment by cket** [comment]
I see now, sorry about that! I've been walking over this and I think I'm getting closer. The job with the EOF error is actually run again despite 'failing completely', for some reason. The job is only run once when I run the pipeline with the forced error failing, so this points to some problem in the restart logic causing a duplicate job being issued.  

More evidence of a scheduling error is that the job is run with the same command and unpickles the file successfully on the second try, whereas the file doesn't exist on the first try. It seems like the failing job was just run too early. My next step is identifying where exactly the job is being issued 2X.

**Comment by chapmanb** [comment]
Christopher -- thanks for the help debugging. The test job I wrote for this should submit the job multiple times:

1. First I manipulate the PATH to make the job fail (by moving `rev`, needed for the job to run, out of the PATH).
2. Then I fix the PATH and rerun the same job with `--restart` so it'll hopefully finish cleanly.
3. Then I run again with `--restart` to test for an expected failure (restarting on a finished job).

I'm not sure if that's relevant to what you're seeing or not but is the logic behind the tests. From the new code side, I'm now calling toil.restart() on the old directory instead of doing a new workflow. Perhaps there should be more tweaking/fixes from my side, I'm just not sure what to do. Thanks again for the help.

**Comment by cket** [comment]
Hi Brad, the problem here seems to be a job being issued multiple times within the second step of your test when the workflow is called with ```--restart```. Since the workflow runs correctly without forcing the initial failure I think this is a ```--restart`` problem, not a workflow or test problem.

When a workflow is running Toil determines what job to run next by looking a finished job's successors. In the ```--restart``` case we are seeing the EOF job being issued despite not being part of any finishing job's successors. The job fails, and is then being issued again (correctly this time) when its predecessor finishes. 

Unfortunately, I'm out of the office this week and the next so I can't dedicate too much more time to debugging but I would look at the restart code to figure out how it determines which jobs are OK to run when restarting the workflow.  Does the CWL support implement its own restart logic? If so, that's where I would looks first.

File: src/toil/test/cwl/cwlTest.py [hunk 1]
```diff
@@ -50,6 +50,35 @@ def test_run_revsort(self):
...
+               os.path.join(cwlDir, "revsort.cwl"), os.path.join(cwlDir, "revsort-job.json")]
+        def path_without_rev():
+            return ":".join([d for d in os.environ["PATH"].split(":")
+                             if not os.path.exists(os.path.join(d, "rev"))])
+        orig_path = os.environ["PATH"]
+        os.environ["PATH"] = path_without_rev()
+        # Force a failure and half finished job
+        try:
+            cwltoil.main(cmd)
+            raise ValueError("Expected problem job did not fail")
```
Comment (hannes-ucsc) [inline src/toil/test/cwl/cwlTest.py]: Use self.fail()

File: src/toil/test/cwl/cwlTest.py [hunk 2]
```diff
@@ -50,6 +50,35 @@ def test_run_revsort(self):
...
+        except AssertionError:
+            pass
```
Comment (hannes-ucsc) [inline src/toil/test/cwl/cwlTest.py]: Why are you swallowing AssertionError here?

File: src/toil/test/cwl/cwlTest.py [hunk 3]
```diff
@@ -50,6 +50,35 @@ def test_run_revsort(self):
...
+        # Finish the job with a correct run
+        os.environ["PATH"] = orig_path
+        cwltoil.main(cmd + ["--restart"])
+        # Should fail because job completed successfully
+        try:
+            cwltoil.main(cmd + ["--restart"])
+            raise AssertionError("Restart with missing directory did not fail")
```
Comment (hannes-ucsc) [inline src/toil/test/cwl/cwlTest.py]: self.fail("…")

File: src/toil/test/cwl/cwlTest.py [hunk 4]
```diff
@@ -50,6 +50,35 @@ def test_run_revsort(self):
...
+        except NoSuchJobStoreException:
+            pass
```
Comment (hannes-ucsc) [inline src/toil/test/cwl/cwlTest.py]: Document why this exception is expected.

File: src/toil/test/cwl/cwlTest.py [hunk 5]
```diff
@@ -50,6 +50,35 @@ def test_run_revsort(self):
...
+        cmd = ['--outdir', outDir, '--jobStore', os.path.join(outDir, 'jobStore'), "--no-container",
+               os.path.join(cwlDir, "revsort.cwl"), os.path.join(cwlDir, "revsort-job.json")]
+        def path_without_rev():
+            return ":".join([d for d in os.environ["PATH"].split(":")
+                             if not os.path.exists(os.path.join(d, "rev"))])
+        orig_path = os.environ["PATH"]
+        os.environ["PATH"] = path_without_rev()
+        # Force a failure and half finished job
+        try:
+            cwltoil.main(cmd)
```
Comment (hannes-ucsc) [inline src/toil/test/cwl/cwlTest.py]: I don't see where the failure is being forced.

Comment (hannes-ucsc) [inline src/toil/test/cwl/cwlTest.py]: Read your PR description just now which answers my question. But I think a comment would help here.

**Comment by hannes-ucsc** [comment]
You might want to pass --logDebug to the test. Maybe that will reveal something.

**Comment by chapmanb** [comment]
Hannes;
Thanks for all the comments. While responding to these I realized the underlying issue might be that cwltoil is exiting poorly for this failure case, and I moved discussion of this to #1122 with more details. I hope if we fix that it'll help here. Would you have any pointers about how best to improve failures with cwltoil runs so they're cleaner?

**Comment by benedictpaten** [comment]
@chapmanb @cket I think this might now work, I fixed the big bug in toil that was preventing it working correctly. Hope this can be unblocked.

**Comment by chapmanb** [comment]
Benedict -- thanks so much for the fixes. This does resolve the problem and get this working correctly. Brilliant. I've updated the tests to synchronize with the latest changes and address Hannes' review and rebased this all into a single clean commit.

So, this looks great to me now and is ready to test/merge if anyone on the Toil team had time to review. Thanks again for all the help getting this working.

**Comment by hannes-ucsc** [comment]
IT failed due to PyPI outage.

**Pull request merged**